content/ui-testing-handbook/react/en/accessibility-testing.md

```diff
@@ -0,0 +1,288 @@
+
+```diff:title=src/components/Task.js
+import React from 'react';
+import PropTypes from 'prop-types';
+import {
+ Checkbox,
+ Flex,
+ IconButton,
+ Input,
+ Box,
+ VisuallyHidden,
+} from '@chakra-ui/react';
+import { StarIcon } from '@chakra-ui/icons';
+
+export const Task = ({
+ task: { id, title, state },
+ onArchiveTask,
+ onTogglePinTask,
+ onEditTitle,
+ ...props
+}) => (
+
+ // code omitted for brevity
+
+   <Box width="full" as="label">
+     <VisuallyHidden>Edit</VisuallyHidden>
+     <Input
+       variant="unstyled"
+       flex="1 1 auto"
+-      color={state === 'TASK_ARCHIVED' ? 'gray.400' : 'gray.700'}
++      color={state === 'TASK_ARCHIVED' ? 'gray.600' : 'gray.700'}
+       textDecoration={state === 'TASK_ARCHIVED' ? 'line-through' : 'none'}
+       fontSize="sm"
+       isTruncated
+       value={title}
+       onChange={(e) => onEditTitle(e.target.value, id)}
+     />
+   </Box>
+
+   // code omitted for brevity
+ </Flex>
+);
+
+Task.propTypes = {
+ task: PropTypes.shape({
+   id: PropTypes.string.isRequired,
+   title: PropTypes.string.isRequired,
+   state: PropTypes.string.isRequired,
+ }),
+ onArchiveTask: PropTypes.func.isRequired,
+ onTogglePinTask: PropTypes.func.isRequired,
+ onEditTitle: PropTypes.func.isRequired,
+};
+```
+
+The second violation, **“Ensures `<li>` elements are used semantically,”** indicates that the DOM structure is incorrect. The Task component renders an `<li>` element. However, it's not wrapped with a `<ul>` in its stories. Which makes sense. These stories are for the Task component. The `<ul>` is actually provided by the TaskList. So the DOM structure gets validated in the TaskList stories. Therefore, it's safe to ignore this error. In fact, we can go ahead and disable this rule for all the Task stories.
+
+```diff:title=src/components/Task.stories.js
+import React from 'react';
+import { Task } from './Task';
+
+export default {
+  component: Task,
+  title: 'Task',
+  argTypes: {
+    onArchiveTask: { action: 'onArchiveTask' },
+    onTogglePinTask: { action: 'onTogglePinTask' },
+    onEditTitle: { action: 'onEditTitle' },
+  },
++  parameters: {
++    a11y: {
++      config: {
++        rules: [{ id: 'listitem', enabled: false }],
++      },
++    },
++  },
+};
+
+// remaining code omitted for brevity
+```
+
+You can now repeat this process for all other components.
+
+Integrating accessibility testing into Storybook streamlines your development workflow. You don’t have to jump between different tools while working on a component. Everything you need is right there in the browser. You can even simulate visual impairments such as deuteranomaly, protanomaly or tritanopia.
+
+![](/ui-testing-handbook/vision-simulator.png)
+
+### Preventing regressions
```

Comment:
I'm always surprised I don't see more usage of [`axeTest` from `@storybook/addon-storyshots-puppeteer`](https://github.com/storybookjs/storybook/tree/next/addons/storyshots/storyshots-puppeteer#axetest), as it does this same Axe check automatically, for every story.

Not a suggestion to change it; just an aside.

Reply:
Totally forgot about that